### PR TITLE
kubeadm: dynamically populate the current/minimum k8s versions

### LINKED ISSUE #1714: Automatic update for version constants
Right now we have three Kubernetes version constants residing in [constants.go](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go). Those are `CurrentKubernetesVersion`, `MinimumControlPlaneVersion` and `MinimumKubeletVersion`.

These need to be bumped for every release cycle and this has already caused us issues in the past (see kubernetes/kubernetes#80834).

We already have the ability to generate a version during our bazel based build, thanks to some Git magic (see the kubeadm [version package](https://github.com/kubernetes/kubernetes/tree/master/cmd/kubeadm/app/version)). This package, however, may not be able to produce a version for builds of source tarballs and some other corner cases.

We need to find a way to generate the version constants, so that we won't need to do manual bumps. But at the same time enable easy development and testing workflow when the kubeadm version package does not provide us with a valid version.

/priority important-longterm
/area releasing
/help

### PR DESCRIPTION
**Pull request by neolit123**
#### What this PR does / why we need it:

Kubeadm requires manual version updates of its current supported k8s
control plane version and minimally supported k8s control plane and
kubelet versions every release cycle.

To avoid that, in constants.go:
- Add the helper function getSkewedKubernetesVersion() that can be
used to retrieve a MAJOR.(MINOR+n).0 version of k8s. It currently
uses the kubeadm version populated in "component-base/version" during
the kubeadm build process.
- Use the function to set existing version constants (variables).

Update util/config/common.go#NormalizeKubernetesVersion() to
tolerate the case where a k8s version in the ClusterConfiguration
is too old for the kubeadm binary to use during code freeze.

Include unit tests for the new utilities.

Update existing unit tests to tolerate the new dynamic versions.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/1714
xref https://github.com/kubernetes/kubeadm/issues/2440

#### Special notes for your reviewer:

**NOTE:** better to review one commit at a time.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
/cc @SataQiu 
i recall that you worked on the original to change this. PTAL if you can for 1.23.

/hold for review

**Review by SataQiu — COMMENTED** [review]
Thanks @neolit123
LGTM!

File: cmd/kubeadm/app/constants/constants_test.go [hunk 1]
```diff
@@ -237,3 +238,61 @@ func TestGetKubernetesServiceCIDR(t *testing.T) {
...
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.isRunningInTestFunc == nil {
+				tc.isRunningInTestFunc = func() bool { return true }
+			}
+			result := getSkewedKubernetesVersionImpl(tc.versionInfo, tc.n, tc.isRunningInTestFunc)
+			if (tc.expectedResult == nil) != (result == nil) {
+				t.Error("expected result to be nil")
```
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/constants/constants_test.go:288]: good point; will update.

[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/constants/constants_test.go]: updated.

**Pull request merged**